src/models/train_model.py

```diff
@@ -0,0 +1,205 @@
+                )
+
+            score = cross_val_score(
+                model, x_train, y_train, n_jobs=-1, cv=cv_splitter, scoring="neg_mean_absolute_error"
+            )
+            cv_mae = score.mean()
+            return cv_mae
+
+        sampler = optuna.samplers.TPESampler(seed=1)
+        study = optuna.create_study(direction="maximize", sampler=sampler)
+        study.optimize(objective, n_trials=self.optim_iters, n_jobs=-1)
+        self.model_best_params = study.best_params
+
+        model = create_model_based_on_best_params(self.model_best_params)
+        model.fit(x_train, y_train)
+        logging.info(f"Model trained with following params: {self.model_best_params}")
+        self.model = model
+        params = str(self.model_best_params)
+        hashed_params = hashlib.md5(params.encode()).hexdigest()
+        today = datetime.today().strftime("%Y%m%d%H%M%S")
+        self.model_name = f"{today}_{hashed_params}.pkl"
+
+        def save_model():
+            logging.info(f"Saving model to the file")
+
+            joblib.dump(self.model, self.model_path / self.model_name)
+            logging.info(f"Saved model as {self.model_name}")
+
+        save_model()
+
+        pred_values = self.model.predict(x_train)
+        train_mae = mean_absolute_error(y_train, pred_values)
+        train_r2 = r2_score(y_train, pred_values)
+        logging.info(f"Train MAE: {round(train_mae, 3)}\nTrain R2: {round(train_r2, 3)}")
+
+    def eval_model(self, x_test, y_test):
+        pred_values = self.model.predict(x_test)
+        test_mae = mean_absolute_error(y_test, pred_values)
+        test_r2 = r2_score(y_test, pred_values)
+        logging.info(f"Test MAE: {round(test_mae, 3)}\nTest R2: {round(test_r2, 3)}")
+
+    def compare_new_model_with_active(self, x_test, y_test):
+        logging.info(f"Comparing new model with the Active one")
+
+        def load_current_active_model():
+            if not os.path.exists(self.model_path / "active"):
+                os.makedirs(self.model_path / "active")
+            for root, dirs, files in os.walk(self.model_path / "active"):
+                if len(files) == 0:
+                    logging.info(f"No active models found")
+                    return None
+                else:
+                    return joblib.load(self.model_path / "active" / files[0])
+
+        def set_model_as_active(model, path, name):
+            # todo: add comparison with current active model
+            if not os.path.exists(path / "active"):
+                os.makedirs(path / "active")
+            for root, dirs, files in os.walk(path / "active"):
+                for f in files:
+                    os.unlink(os.path.join(root, f))
+                for d in dirs:
+                    shutil.rmtree(os.path.join(root, d))
+            logging.info(f"Making new model as active")
+            joblib.dump(model, path / "active" / name)
+
+        active_model = load_current_active_model()
```

Comment:
again, function without arguments, okey to load but until going into definition, params are not known